tests/test_default_values.py

```diff
@@ -0,0 +1,108 @@
+the conditions stated in the LICENSE file in the project root for
+details.
+
+The above copyright notice and this permission notice shall be included
+in all copies or substantial portions of the Software.
+"""
+
+import unittest
+
+import configsuite
+from configsuite import MetaKeys as MK
+from configsuite import types
+from .data import car
+
+
+class TestDefaultValues(unittest.TestCase):
+    def test_default_values(self):
+        raw_config = car.build_all_default_config()
+        car_schema = car.build_schema()
+        config_suite = configsuite.ConfigSuite(raw_config, car_schema)
+
+        self.assertTrue(config_suite.valid)
+
+        # Unspecified containers are defaulted to empty container of their type
+        self.assertEqual(config_suite.snapshot.owner, {})
+        self.assertEqual(config_suite.snapshot.color, {})
+        self.assertEqual(config_suite.snapshot.minor_incidents, [])
+
+    def test_default_presedence(self):
+        raw_config = car.build_all_default_config()
+        car_schema = car.build_schema()
+        raw_config["tire"] = {}
+        raw_config["tire"]["dimension"] = 15
+        config_suite = configsuite.ConfigSuite(raw_config, car_schema)
+
+        self.assertTrue(config_suite.valid)
+        self.assertEqual(config_suite.snapshot.tire.rim, "green")
+        self.assertEqual(config_suite.snapshot.tire.dimension, 15)
+
+    def test_partial_default_values(self):
+        raw_config = car.build_all_default_config()
+        car_schema = car.build_schema()
+        car_schema[MK.Content]["tire"][MK.Content]["dimension"].pop("default")
+
+        raw_config["tire"] = {}
+        raw_config["tire"]["dimension"] = 15
+        config_suite = configsuite.ConfigSuite(raw_config, car_schema)
+
+        self.assertTrue(config_suite.valid)
+        self.assertEqual(config_suite.snapshot.tire.rim, "green")
+        self.assertEqual(config_suite.snapshot.tire.dimension, 15)
+
+    def test_invalid_defaults(self):
+        car_schema = car.build_schema()
+        raw_config = car.build_all_default_config()
+        configsuite.ConfigSuite(raw_config, car_schema)
+
+        car_schema[MK.Content]["tire"][MK.Content]["dimension"][
+            MK.Default
+        ] = "incorrect default type"
+        with self.assertRaises(ValueError):
+            configsuite.ConfigSuite(raw_config, car_schema)
+            # self.assertTrue(msg.exception.startswidth("default problems!"))
+
+        car_schema = car.build_schema()
+        # List does not allow default
+        car_schema[MK.Content]["minor_incidents"].update(
+            {MK.Default: "correct default type"}
+        )
+        with self.assertRaises(ValueError):
+            configsuite.ConfigSuite(raw_config, car_schema)
+            # self.assertTrue(msg.exception.startswidth("default problems!"))
+
+        car_schema = car.build_schema()
+        # Dict does not allow default
+        car_schema[MK.Content]["owner"][MK.Content][MK.Key][
```

Comment:
Are not this and the line below doing the exact same thing? Also, I suggest we use a somewhat reasonable default value instead, like `{"default_key": "default_value"}`.

Reply:
When I look at this, does not the section below test exactly this and hence this block being a bit mute?

Reply:
Also, please make a similar test for `list`.

Reply:
And a test that tries to specify a default value for a `BasicType` inside a list or dict, which should not be allowed.

Reply:
This is what I want to test:
Given the setup:
```
"owner": {
                MK.Type: types.Dict,
                MK.Required: False,
                MK.Content: {
                    MK.Key: {MK.Type: types.String},
                    MK.Value: {MK.Type: types.Number},
                },
```
I want to update the key and value with default values (which is not allowed: for dicts)

```
"owner": {
                MK.Type: types.Dict,
                MK.Required: False,
                MK.Content: {
                    MK.Key: {MK.Type: types.String, MK.Default="correct default type"},
                    MK.Value: {MK.Type: types.Number, MK.Default=10},
                },
```
In that case the lines are updating key and value separately.

The list test is tested before the dict.

These tests does indeed set the values to `BasicType` for each of list and dict.

Afterwards I'm setting up a test where the `MK.Default` is set on the `NamedDict` itself
```
# Default only valid at basic types
        schema_value_content = {
            MK.Type: types.NamedDict,
            MK.Default: {"profile": 200, "width": 20},
            MK.Content: {
                "profile": {MK.Type: types.Number},
                "width": {MK.Type: types.Number},
            },
        }
```